vispy/visuals/graphs/layouts/force_directed.py

```diff
@@ -0,0 +1,212 @@
+        layout : tuple
+            For each iteration of the layout calculation it yields a tuple
+            containing (node_vertices, line_vertices, arrow_vertices). These
+            vertices can be passed to the `MarkersVisual` and `ArrowVisual`.
+        """
+        if adjacency_mat.shape[0] != adjacency_mat.shape[1]:
+            raise ValueError("Adjacency matrix should be square.")
+
+        self.num_nodes = adjacency_mat.shape[0]
+
+        if self.num_nodes < 500:
+            # Use the sparse solver
+            solver = self._sparse_fruchterman_reingold
+        else:
+            solver = self._fruchterman_reingold
+
+        for result in solver(adjacency_mat, directed):
+            yield result
+
+    def _fruchterman_reingold(self, adjacency_mat, directed=False):
+        if self.optimal is None:
+            self.optimal = 1 / np.sqrt(self.num_nodes)
+
+        if self.pos is None:
+            # Random initial positions
+            pos = np.asarray(
+                np.random.random((self.num_nodes, self.dim)),
+                dtype='f32'
+            )
+        else:
+            pos = self.pos.astype('f32')
+
+        # Yield initial positions
+        line_vertices, arrows = straight_line_vertices(adjacency_mat, pos,
+                                                       directed)
+        yield pos, line_vertices, arrows
+
+        # The initial "temperature"  is about .1 of domain area (=1x1)
+        # this is the largest step allowed in the dynamics.
+        t = 0.1
+
+        # Simple cooling scheme.
+        # Linearly step down by dt on each iteration so last iteration is
+        # size dt.
+        dt = t / float(self.iterations+1)
+        delta = np.zeros(
+            (pos.shape[0], pos.shape[0], pos.shape[1]),
+            dtype=adjacency_mat.dtype
+        )
+
+        # The inscrutable (but fast) version
+        # This is still O(V^2)
+        # Could use multilevel methods to speed this up significantly
+        for iteration in range(self.iterations):
+            # Matrix of difference between points
+            for i in range(pos.shape[1]):
+                delta[:, :, i] = pos[:, i, None] - pos[:, i]
+
+            # Distance between points
+            distance = np.sqrt((delta**2).sum(axis=-1))
+            # Enforce minimum distance of 0.01
+            distance = np.where(distance < 0.01, 0.01, distance)
+            # Displacement "force"
+            displacement = np.transpose(
+                np.transpose(delta) * (
+                    self.optimal * self.optimal/distance**2 -
+                    adjacency_mat*distance/self.optimal
+                )
+            ).sum(axis=1)
+
+            # update positions
+            length = np.sqrt((displacement**2).sum(axis=1))
+            length = np.where(length < 0.01, 0.1, length)
+            delta_pos = np.transpose(np.transpose(displacement)*t/length)
+            pos += delta_pos
+            pos = rescale_layout(pos)
```

Comment:
I think this should be changed somehow, but I'm not sure yet how: It normalizes the node coordinates, but each iteration again. This modifies the current iteration values which may not be desirable (networkx only normalizes them at the end)

If anyone knows some other way to make sure the positions stay within [0, 1], please let me know.